### PR TITLE
Benchmark online

### PR DESCRIPTION
**Pull request by MayMeta**
based on `benchmark.py`

### CODE REVIEW

**Comment by MayMeta** [comment]
```
EXAMPLE:
export CUDA_VISIBLE_DEVICES=1
python benchmark_online.py \
--checkpoint best_checkpoints/JasperNetBig_NovoGrad_lr1e-4_wd1e-3_bs512___ \
finetune_after_self_train_on_2020_11_27_clean_valset_epoch144_iter0335000.pt \
--device cuda -T 6.0 --benchmark-duration 60 --rps 5.0

OUT:
initializing model...
batch [1, 48000] | audio 6.00 sec
Warming up for 100 iterations...
Warmup done in 2.7 sec
Starting 60 second benchmark (300 requests, rps 5.0)...
avg gap between requests: 195.0 ms

Latency mean: 33.6 ms, median: 33.1 ms, 90-th percentile: 36.9 ms, 95-th percentile: 46.2 ms,
99-th percentile: 60.7 ms, max: 73.0 ms | service idle time fraction: 81.7%
```

**Comment by MayMeta** [comment]
todo: run on onnx

**Comment by II245** [comment]
lgtm

File: benchmark_online.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+	if use_cuda:
+		print(f'CUDA_VISIBLE_DEVICES: {os.environ.get("CUDA_VISIBLE_DEVICES")!r}')
+	print('initializing model...')
+
+	if args.onnx:
+		# todo: pass dict with provider setting when we will migrate to onnxruntime>=1.7
+		onnxruntime_session = onnxruntime.InferenceSession(args.onnx)
+		model = lambda x: onnxruntime_session.run(None, dict(x=x, xlen=[1.0] * len(x)))
+		load_batch = lambda x: x.numpy()
+		args.sample_rate = 8000
```
Comment (vadimkantorov) [inline benchmark_online.py]: why do you reset it to 8000? if only 8000 is supported please `assert args.sample_rate == 8000` instead. we don't do these asserts in other places, so removing this should be fine as well

if you wish to do this properly, please figure out how to save meta information in the onnx file (in train.py there are some pointers in comments on how to do that)

Comment (MayMeta) [reply benchmark_online.py]: Please, advise on how to extract `sample_rate` from `onnxruntime_session`

File: benchmark_online.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	if use_cuda:
+		torch.backends.cudnn.benchmark = True
+	torch.set_grad_enabled(False)  # no back-prop - no gradients
+	for i in range(args.warmup_iterations):
+		model(load_batch(batch))
+	print(f'Warmup done in {tictoc() - tic_wall:.1f} sec\n')
+
+	n_requests = int(round(args.benchmark_duration * args.rps))
+	print(f'Starting {args.benchmark_duration} second benchmark ({n_requests} requests, rps {args.rps:.1f})...')
+	schedule = np.random.random(n_requests) * args.benchmark_duration  # uniform random distribution of requests
```
Comment (vadimkantorov) [inline benchmark_online.py]: if these np.random.random, np.sort and np.diff, np.quantile are all numpy usages, please replace them with torch equivalents

File: benchmark_online.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	print(f'avg gap between requests: {np.diff(schedule).mean() * 1e3:.1f} ms')
+	latency_times, idle_times = [], []
+	slow_warning = False
+	for t_request in schedule:
+		tic = tictoc()
+		if tic < t_request:  # no requests yet. at this point prod would wait for the next request
+			sleep_time = t_request - tic
+			idle_times.append(sleep_time)
+			time.sleep(sleep_time)
+		elif tic > t_request + 0.5 and not slow_warning:
```
Comment (vadimkantorov) [inline benchmark_online.py]: what's the meaning of 0.5 constant here? please move it to ArgumentParser with some meaningful name

File: benchmark_online.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+			checkpoint['args'].get(key) for key in
+			['model', 'lang', 'sample_rate', 'window_size', 'window_stride', 'window', 'num_input_features']
+		)
+
+		text_config = json.load(open(args.text_config))
+		text_pipelines = []
+		for pipeline_name in args.text_pipelines:
+			text_pipelines.append(text_processing.ProcessingPipeline.make(text_config, pipeline_name))
+
+		frontend = models.LogFilterBankFrontend(
```
Comment (vadimkantorov) [inline benchmark_online.py:85]: please use kwargs for all arguments. Othewise, the code may become silently wrong if we reorder arguments

**Comment by vadimkantorov** [comment]
looks quite good. please fix the minor comments, and then let's merge this

File: benchmark_online.py (hunk 1 above)
Comment (MayMeta) [reply benchmark_online.py]: Done. Moved to `--sample-rate`.

File: benchmark_online.py (hunk 2 above)
Comment (MayMeta) [reply benchmark_online.py]: Done

File: benchmark_online.py (hunk 3 above)
Comment (MayMeta) [reply benchmark_online.py]: Done. Added `--max-latency` parameter

File: benchmark_online.py (hunk 4 above)
Comment (MayMeta) [reply benchmark_online.py:85]: Done

**Pull request merged**